File: python/leetcode/1046-last-stone-weight.py
import heapq
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        stones = [stone *-1 for stone in stones]
        heapq.heapify(stones)
        
        while len(stones) > 1:
            y = heapq.heappop(stones) * -1
            x = heapq.heappop(stones) * -1
            
            if x != y:
                heapq.heappush(stones, (y-x) * -1)
            
        return stones[0] * -1 if len(stones) > 0 else 0

File: c#/leetcode/1046-last-stone-weight.cs
public class Solution {
    public int LastStoneWeight(int[] stones) {
        var q = new PriorityQueue<int, int>(Comparer<int>.Create((x,y) => y - x));
        
        foreach (var stone in stones) {
            q.Enqueue(stone, stone);
        }
        
        while (q.Count >= 2) {
            int y = q.Dequeue();
            int x = q.Dequeue();
            
            if (x != y) {
                q.Enqueue((y-x), (y-x));
                
            }
        }
        
        return q.Count == 1 ? q.Peek() : 0;
    }
}